be/src/connector/binlog_connector.h

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "column/column.h"
+#include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
+#include "connector/connector.h"
+#include "storage/tablet.h"
+
+namespace starrocks::connector {
+
+class BinlogConnector final : public Connector {
+public:
+    ~BinlogConnector() override = default;
+
+    DataSourceProviderPtr create_data_source_provider(ConnectorScanNode* scan_node,
+                                                      const TPlanNode& plan_node) const override;
+
+    ConnectorType connector_type() const override { return ConnectorType::BINLOG; }
+};
+
+class BinlogDataSource;
+
+class BinlogDataSourceProvider final : public DataSourceProvider {
+public:
+    ~BinlogDataSourceProvider() override = default;
+    friend class BinlogDataSource;
+    BinlogDataSourceProvider(ConnectorScanNode* scan_node, const TPlanNode& plan_node);
+    DataSourcePtr create_data_source(const TScanRange& scan_range) override;
+
+    bool insert_local_exchange_operator() const override { return false; }
+    bool accept_empty_scan_ranges() const override { return false; }
+
+    bool stream_data_source() const override { return true; }
+
+protected:
+    ConnectorScanNode* _scan_node;
+    const TBinlogScanNode _binlog_scan_node;
+};
+
+using BinlogMetaFieldMap = std::unordered_map<std::string, VectorizedFieldPtr>;
+const std::string BINLOG_PREFIX = "_binlog";
+const std::string BINLOG_OP = BINLOG_PREFIX + "_op";
+const std::string BINLOG_VERSION = BINLOG_PREFIX + "_version";
+const std::string BINLOG_SEQ_ID = BINLOG_PREFIX + "_seq_id";
+const std::string BINLOG_TIMESTAMP = BINLOG_PREFIX + "_timestamp";
+
+class BinlogDataSource final : public DataSource {
+public:
+    ~BinlogDataSource() override = default;
+
+    BinlogDataSource(const BinlogDataSourceProvider* provider, const TScanRange& scan_range);
+    Status open(RuntimeState* state) override;
+    void close(RuntimeState* state) override;
+    Status get_next(RuntimeState* state, ChunkPtr* chunk) override;
+
+    int64_t raw_rows_read() const override;
```

Comment:
It's too trivial, these several metrics are used for audit-log and resourcegroup scheduling, IMO only resource-gropu scheduling related metrics are necessary.

Reply:
Maybe I do not get your point. These functions are pure virtual in `DataSource`, and do you mean there is no need to implement them?